hibernate-osgi/src/main/java/org/hibernate/osgi/OsgiClassLoader.java

```diff
@@ -8,33 +8,41 @@
 
 import java.net.URL;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.Enumeration;
 import java.util.HashMap;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.NoSuchElementException;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 
 import org.hibernate.service.spi.Stoppable;
 import org.osgi.framework.Bundle;
 
 /**
  * Custom OSGI ClassLoader helper which knows all the "interesting"
  * class loaders and bundles.  Encapsulates the OSGi related CL capabilities.
- * 
+ *
  * @author Brett Meyer
  * @author Tim Ward
  */
 public class OsgiClassLoader extends ClassLoader implements Stoppable {
 	// Leave these as Sets -- addClassLoader or addBundle may be called more
 	// than once if a SF or EMF is closed and re-created.
-	private Set<ClassLoader> classLoaders = new LinkedHashSet<ClassLoader>();
-	private Set<Bundle> bundles = new LinkedHashSet<Bundle>();
+	// HHH-12553: must be thread-safe. Concurrent impl. would be best, but we have to retain insertion-order.
+	private Set<ClassLoader> classLoaders = Collections.synchronizedSet(new LinkedHashSet<ClassLoader>());
+	private Set<Bundle> bundles = Collections.synchronizedSet(new LinkedHashSet<Bundle>());
+
+	private ConcurrentMap<String, Class<?>> classCache = new ConcurrentHashMap<String, Class<?>>();
+	private ConcurrentMap<String, URL> resourceCache = new ConcurrentHashMap<String, URL>();
+
+	static {
+		ClassLoader.registerAsParallelCapable();
```

Comment:
We might consider leaving this out, for now.  Honestly not sure if there are implications for `registerAsParallelCapable` in OSGi environments, nor am I sure if most OSGi containers even implement this.  @timothyjward, happen to know off the top of your head?

Reply:
Depends on the ClassLoaders in use - JBOsgi's CL e.g. does implement it (KAraf as well afaik).
If in other containers the CL for the OSGi-bundles does not implement it, it becomes ineffective anyway.
Also fine with leaving it out.

Reply:
Apache Felix certainly does - see https://github.com/apache/felix/blob/1878353ae257681e72e50c1c87f3900a13baed4e/framework/src/main/java/org/apache/felix/framework/BundleWiringImpl.java#L1982

Reply:
@timothyjward Bonus points for the super fast answer ;)
@MrEasy We primarily target Felix and Equinox, and it looks like Equinox is supporting this as well.

Reply:
Hi, not an OSGi user myself so I can only guess what `registerAsParallelCapable` means.. no objections to include it but it would be appreciate for such enhancements to have their separate JIRA and separeate PRs.

It would help with our flow and also help PRs being merged faster.

Reply:
+1 -- I doubt it affects much in this specific case, but the separation of commits gives more traceability if side effects kick in.